### ISSUE TITLE
`screenstring()` can show not up to date cell value

### ISSUE THREAD

**Issue by echasnovski** [neovim/neovim#25644]
### Problem

Calling `screenstring()` can return not up to date information.

### Steps to reproduce

1. Create the following config file:
```lua
_G.test_screenstring = function()
  local buf_id = vim.api.nvim_create_buf(false, true)
  vim.api.nvim_buf_set_lines(buf_id, 0, -1, false, { string.rep('a', 1000) })
  local config = { relative = 'editor', row = 1, col = 1, height = 10, width = 10 }
  local win_id = vim.api.nvim_open_win(buf_id, false, config)

  _G.cell_55 = {}

  _G.cell_55.immediate = vim.fn.screenstring(5, 5)
  vim.schedule(function() _G.cell_55.scheduled = vim.fn.screenstring(5, 5) end)
  vim.defer_fn(function() _G.cell_55.deferred = vim.fn.screenstring(5, 5) end, 10)
end
```
2. Call `nvim` with the config file.
3. Execute `:lua _G.test_screenstring()`.
4. Execute `:=cell_55`.

### Expected behavior

The `cell_55` table contains only `"a"` as values, i.e. it looks like `{ deferred = "a", immediate = "a", scheduled = "a" }`.

However, its values varies:
- As far back as 0.8.3 its values are `{ deferred = "a", immediate = " ", scheduled = "a" }`.
- On `bf70a33` its values are `{ deferred = "a", immediate = " ", scheduled = " " }` while on `ebe489d` it is the same as in previous step. This seems to be the result of #25629. Applying the following patch makes it behave as was previously:
```patch
diff --git a/src/nvim/getchar.c b/src/nvim/getchar.c
index 5f3b14399..46a786c1d 100644
--- a/src/nvim/getchar.c
+++ b/src/nvim/getchar.c
@@ -2862,9 +2862,7 @@ int inchar(uint8_t *buf, int maxlen, long wait_time)

     // Always flush the output characters when getting input characters
     // from the user and not just peeking.
-    if (wait_time == -1L || wait_time > 10L) {
-      ui_flush();
-    }
+    ui_flush();

     // Fill up to a third of the buffer, because each character may be
     // tripled below.
```

### Neovim version (nvim -v)

NVIM v0.10.0-dev-bf70a33

### Vim (not Nvim) behaves the same?

No (doesn't have this functionality)

### Operating system/version

EndeavourOS Linux x86_64 (6.5.6-arch1-1)

### Terminal name/version

st-0.9

### $TERM environment variable

st-256color

### Installation

appimage

**Comment by bfredl** [comment]
`immediate = "a"` would imply an implicit redraw as part of `screenstring()`. I don't think we want that. Rather a doc improvement, i e expand "This is mainly to be used for testing." with _how_ the function should be used for testing.

**Comment by echasnovski** [comment]
> `immediate = "a"` would imply an implicit redraw as part of `screenstring()`. I don't think we want that. Rather a doc improvement, i e expand "This is mainly to be used for testing." with _how_ the function should be used for testing.

That's reasonable. It worked for testing for me thus far, but bf70a33 seems to be the culprit of some recent problems.

**Comment by zeertzjq** [comment]
Not a bug. The docs say "on screen", and the floating window isn't drawn on screen yet when `screenstring()` is called. Use `:redraw` to make sure pending screen updates are drawn.

**Comment by echasnovski** [comment]
That's... unfortunate. I would argue that drawing should not be delayed so much in unblocked state that it needs extra command to appear.

**Comment by zeertzjq** [comment]
Also, if the command is changed to `:lua vim.schedule(_G.test_screenstring)`, then same behavior happens in earlier versions.
